### DISCUSSION TITLE
Proposal: Generics in Attributes

### DISCUSSION THREAD

**Discussion by ChristopherHaas** [dotnet/csharplang#4782] [Language Ideas] [answered]
In C#, attributes are often used to declare a relationship to a type. Today, we cannot use generics for this; thus, we end up with attributes like the following example and rely on runtime checking that the type is valid.

`[Proxy(typeof(ObjectProxy))]`

Instead, it would be awesome if we could have something like this, and compile-time enforce constraints on the generic type.

`[Proxy<ObjectProxy>]`

Furthermore, it would be very useful if we could use the type on which the attribute is declared in the attribute type definition. I.e.

```
[Proxy]
class Test

class ProxyAttribute<TDeclarer> : Attribute
```

where TDeclarer would be _Test_

Similar for properties:

```
class Test {

[Proxy]
string Property { get; }
}

class ProxyAttribute<TDeclarer> : Attribute
```

where TDeclarer would be _string_.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/124.

**Reply by BreyerW** [reply]
I hope this will finally make in with c# 10 or 11 because with abstract static members in interfaces it will be possible to do some interesting stuff that cant be done with typeof (at least not easily). In the past it wasnt particularly important since almost everything could be done with typeof workaround but abstract static changes situation a bit

**Reply by RikkiGibson** [reply]
Could you please elaborate on what scenarios using abstract statics would be improved by allowing generics in attributes?

**Reply by BreyerW** [reply]
Eg factory pattern without having to go through reflection via abstract statics.

Generally you can achieve everything through typeof and reflection but it will be slow or unpleasant, error prone/very hard to get right (since reflection wasnt updated in ages and as such it drags behind runtime capabilities)  or everything at the same time
